zerver/tests/test_auth_backends.py

```diff
@@ -404,419 +405,464 @@ def json(self) -> str:
+            """If the user doesn't exist yet in closed realm, give an error"""
+            email = "nonexisting@phantom.com"
+            name = 'Full Name'
+            account_data_dict = self.get_account_data_dict(email=email, name=name)
+            result = self.social_auth_test(account_data_dict,
+                                           subdomain='zulip')
+            self.assertEqual(result.status_code, 302)
+            data = load_subdomain_token(result)
+            self.assertEqual(data['email'], email)
+            self.assertEqual(data['name'], name)
+            self.assertEqual(data['subdomain'], 'zulip')
+            self.assertEqual(result.status_code, 302)
+            parsed_url = urllib.parse.urlparse(result.url)
+            uri = "{}://{}{}".format(parsed_url.scheme, parsed_url.netloc,
+                                     parsed_url.path)
+            self.assertTrue(uri.startswith('http://zulip.testserver/accounts/login/subdomain/'))
+
+            result = self.client_get(result.url)
+            self.assertEqual(result.status_code, 200)
+            self.assert_in_response('action="/register/"', result)
+            self.assert_in_response('Your email address, {}, is not '
+                                    'in one of the domains that are allowed to register '
+                                    'for accounts in this organization.'.format(email), result)
+
+        def test_social_auth_complete(self) -> None:
+            with mock.patch('social_core.backends.oauth.BaseOAuth2.process_error',
+                            side_effect=AuthFailed('Not found')):
+                result = self.client_get(reverse('social:complete', args=[self.backend.name]))
+                self.assertEqual(result.status_code, 302)
+                self.assertIn('login', result.url)
+
+        def test_social_auth_complete_when_base_exc_is_raised(self) -> None:
+            with mock.patch('social_core.backends.oauth.BaseOAuth2.auth_complete',
+                            side_effect=AuthStateForbidden('State forbidden')), \
+                    mock.patch('zproject.backends.logging.warning'):
+                result = self.client_get(reverse('social:complete', args=[self.backend.name]))
+                self.assertEqual(result.status_code, 302)
+                self.assertIn('login', result.url)
+
+class GitHubAuthBackendTest(SocialAuthBaseWrapper.SocialAuthBase):
+    CLIENT_KEY_SETTING = "SOCIAL_AUTH_GITHUB_KEY"
+
+    def setUpHook(self) -> None:
+        self.login_url = "/accounts/login/social/github"
+        self.signup_url = "/accounts/register/social/github"
+        self.authorization_url = "https://github.com/login/oauth/authorize"
+        self.access_token_url = "https://github.com/login/oauth/access_token"
+        self.user_info_url = "https://api.github.com/user"
+        self.auth_finish_url = "/complete/github/"
+        self.config_error_url = "/config-error/github"
```

Comment:
I think all of these URL properties may belong as values on the class rather than set by the setUp hook.

If we did that migration, and made `BACKEND_CLASS = GitHubAuthBackend`, then I think you might be able to get rid of `setUpHook` entirely and just have it be a set of variables on the base class.  i.e. the code in the `setUp` method would be just have these added to the end:

```
self.backend = self.BACKEND_CLASS()
self.backend.strategy = DjangoStrategy(storage=BaseDjangoStorage())
self.user_profile.backend = self.backend
```